collector/btrfs_linux.go

```diff
@@ -0,0 +1,189 @@
+	const subsystem = "btrfs"
+
+	// Basic information about the filesystem.
+	devLabels := []string{"uuid"}
+
+	// Retrieve the metrics.
+	metrics := c.getMetrics(s)
+
+	// Convert all gathered metrics to Prometheus Metrics and add to channel.
+	for _, m := range metrics {
+		labels := append(devLabels, m.extraLabel...)
+
+		desc := prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, subsystem, m.name),
+			m.desc,
+			labels,
+			nil,
+		)
+
+		labelValues := []string{s.UUID}
+		if len(m.extraLabelValue) > 0 {
+			labelValues = append(labelValues, m.extraLabelValue...)
+		}
+
+		ch <- prometheus.MustNewConstMetric(
+			desc,
+			prometheus.GaugeValue,
+			m.value,
+			labelValues...,
+		)
+	}
+}
+
+// getMetrics returns metrics for the given Btrfs statistics.
+func (c *btrfsCollector) getMetrics(s *btrfs.Stats) []btrfsMetric {
+	metrics := []btrfsMetric{
+		{
+			name:            "info",
+			desc:            "Filesystem information",
+			value:           1,
+			extraLabel:      []string{"label"},
+			extraLabelValue: []string{s.Label},
+		},
+		{
+			name:  "global_rsv_size_bytes",
+			desc:  "Size of global reserve.",
+			value: float64(s.Allocation.GlobalRsvSize),
+		},
+	}
+
+	// Information about devices.
+	for n, dev := range s.Devices {
+		metrics = append(metrics, btrfsMetric{
+			name:            "device_size_bytes",
+			desc:            "Size of a device that is part of the filesystem.",
+			value:           float64(dev.Size),
+			extraLabel:      []string{"device"},
+			extraLabelValue: []string{n},
+		})
+	}
+
+	// Information about data, metadata and system data.
+	metrics = append(metrics, c.getAllocationStats("data", s.Allocation.Data)...)
+	metrics = append(metrics, c.getAllocationStats("metadata", s.Allocation.Metadata)...)
+	metrics = append(metrics, c.getAllocationStats("system", s.Allocation.System)...)
+
+	return metrics
+}
+
+// getAllocationStats returns allocation metrics for the given Btrfs Allocation statistics.
+func (c *btrfsCollector) getAllocationStats(a string, s *btrfs.AllocationStats) []btrfsMetric {
+	metrics := []btrfsMetric{
+		{
+			name:            "reserved_bytes",
+			desc:            "Amount of space reserved for a data type",
+			value:           float64(s.ReservedBytes),
+			extraLabel:      []string{"block_group_type"},
+			extraLabelValue: []string{a},
+		},
+	}
+
+	// Add all layout statistics.
+	for layout, stats := range s.Layouts {
+		metrics = append(metrics, c.getLayoutStats(a, layout, stats)...)
+	}
+
+	return metrics
+}
+
+// getLayoutStats returns metrics for a data layout.
+func (c *btrfsCollector) getLayoutStats(a, l string, s *btrfs.LayoutUsage) []btrfsMetric {
+	return []btrfsMetric{
+		{
+			name:            "used_bytes",
+			desc:            "Amount of used space by a layout/data type",
+			value:           float64(s.UsedBytes),
+			extraLabel:      []string{"block_group_type", "mode"},
+			extraLabelValue: []string{a, l},
+		},
+		{
+			name:            "total_bytes",
```

Comment:
I think we call this `size_bytes` in the standard filesystem metrics.

Reply:
This is different than the filesystem size: Btrfs allocates chunks on the fly, and this indicates how much data is currently allocated. In the example below it is indicated using `Size` (as it is the current size of that block group type and allocation mode). Following this name is probably more clear for users.

```
Overall:
    Device size:		 291.16GiB
    Device allocated:		 242.03GiB
    Device unallocated:		  49.13GiB
    Device missing:		 291.16GiB
    Used:			 236.34GiB
    Free (estimated):		  54.53GiB	(min: 54.53GiB)
    Data ratio:			      1.00
    Metadata ratio:		      1.00
    Global reserve:		 512.00MiB	(used: 0.00B)

Data,single: Size:236.00GiB, Used:230.59GiB (97.71%)

Metadata,single: Size:6.00GiB, Used:5.75GiB (95.82%)

System,single: Size:32.00MiB, Used:48.00KiB (0.15%)
```

Reply:
That really reinforces the idea that this should be `size`, not `total`.